syntax = "proto3";

package chat;

service ChatService {
  rpc Login(LoginRequest) returns (Response);
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse);
  rpc SendMessage(SendMessageRequest) returns (Response);
  rpc ReadMessages(ReadMessagesRequest) returns (ReadMessagesResponse);
  rpc DeleteMessage(DeleteMessageRequest) returns (Response);
  rpc DeleteAccount(DeleteAccountRequest) returns (Response);
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message ListAccountsRequest {
  int32 page_num = 1;
}

message ListAccountsResponse {
  repeated string usernames = 1;
}

message SendMessageRequest {
  string recipient = 1;
  string message = 2;
}

message ReadMessagesRequest {
  string username = 1;
  int32 limit = 2;
}

message ReadMessagesResponse {
  repeated Message messages = 1;
}

message Message {
  int32 id = 1;
  string sender = 2;
  string message = 3;
}

message DeleteMessageRequest {
  int32 message_id = 1;
}

message DeleteAccountRequest {
  string username = 1;
  string password = 2;
}

message Response {
  string status = 1;
  string message = 2;
}
